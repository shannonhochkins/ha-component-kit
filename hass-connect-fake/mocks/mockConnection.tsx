import type { HassEntities } from "home-assistant-js-websocket";
import 'jest-localstorage-mock';
import { entities } from './mockEntities';
const timeRegex = /\d{2}:\d{2}/g;
const dateRegex = /20(2[3-9]|[3-9][0-9])-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z/g;
const ENTITIES = JSON.parse(JSON.stringify(entities)  
  // purposely using 2022 here as the regex should work for 9 years ;)
  .replace(dateRegex, '2022-07-23T03:14:01.164Z')
  .replace(timeRegex, '01:01'));

localStorage.setItem('hassTokens', JSON.stringify({
  access_token: "FAKE",
  token_type:"Bearer",
  expires_in: Number.MAX_SAFE_INTEGER,
  hassUrl: "http://fake.com",
  clientId: "http://localhost:6006/",
  expires: Number.MAX_SAFE_INTEGER,
  refresh_token:"FAKE"
}));
// eslint-disable-next-line react-refresh/only-export-components
export const connection = {
  id: 'fake'
}
// eslint-disable-next-line react-refresh/only-export-components
export const mocked = {
  callService: jest.fn(),
  getAllEntities: jest.fn().mockImplementation(() => ENTITIES),
  async getAuth() {
    return {
      expired: false,
    }
  },
  async getConfig() {
    return {
      time_zone: 'Australia/Sydney'
    }
  },
  async createConnection() {
    return connection;
  },
  subscribeEntities: jest.fn().mockImplementation((_: null, subScribeFn: (entities: HassEntities) => void) => {
    subScribeFn(ENTITIES);
  })
}

jest.mock('home-assistant-js-websocket', () => mocked);

import { HassConnect } from '@hakit/core';

// eslint-disable-next-line react-refresh/only-export-components
export const onReady = jest.fn();
// Custom wrapper component for testing
export const TestWrapper = ({ children }: { children: React.ReactNode }) => {
  return <HassConnect onReady={onReady} hassUrl="http://fake.com">
    {children}
  </HassConnect>
}