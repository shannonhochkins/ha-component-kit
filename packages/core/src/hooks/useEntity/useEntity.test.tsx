import { TestWrapper, onReady, mocked, connection } from "@mocks/mockConnection";
import { renderHook, act, waitFor } from "@testing-library/react";
import { useEntity } from "@core";
import { omit } from "lodash";

describe("useEntity", () => {
  beforeEach(() => {
    onReady.mockClear();
  });
  it("should allow the entity to call a service directly from the entity object", async () => {
    const { result } = renderHook(() => useEntity("light.fake_light_1"), {
      wrapper: TestWrapper,
    });
    await waitFor(() => expect(onReady).toHaveBeenCalledTimes(1));
    act(() => {
      result.current.service.turnOn({
        color_name: "red",
      });
    });

    expect(mocked.callService).toHaveBeenCalledWith(
      connection,
      "light",
      "turn_on",
      {
        color_name: "red",
      },
      {
        entity_id: "light.fake_light_1",
      },
    );
  });

  it("should retrieve the entity object", async () => {
    const { result } = renderHook(() => useEntity("light.fake_light_1"), {
      wrapper: TestWrapper,
    });
    await waitFor(() => expect(onReady).toHaveBeenCalledTimes(1));
    act(() => {
      // omitting the api object here as the serializer doesn't like proxy objects
      expect(omit(result.current, "api", "last_changed", "last_updated")).toMatchSnapshot();
    });
  });

  it("should throw an error when provided with a entity name that does not exist", async () => {
    await act(async () => {
      await renderHook(
        async () => {
          await waitFor(() => expect(onReady).toHaveBeenCalledTimes(1));
          expect(useEntity("light.does_not_exist")).toThrowError("Entity light.does_not_exist not found");
        },
        {
          wrapper: TestWrapper,
        },
      );
    });
  });

  it("should return null instead of throwing an error if returnNullIfNotFound is true", async () => {
    await act(async () => {
      await renderHook(
        async () => {
          await waitFor(() => expect(onReady).toHaveBeenCalledTimes(1));
          expect(
            useEntity("light.does_not_exist", {
              returnNullIfNotFound: true,
            }),
          ).toEqual(null);
        },
        {
          wrapper: TestWrapper,
        },
      );
    });
  });
});
