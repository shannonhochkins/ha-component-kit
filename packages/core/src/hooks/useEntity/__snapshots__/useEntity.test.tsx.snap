// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useEntity should retrieve the entity object 1`] = `
{
  "attributes": {
    "brightness": 255,
    "color_mode": "hs",
    "effect_list": [
      "Night",
      "Read",
      "Meeting",
      "Leasure",
      "Soft",
      "Rainbow",
      "Shine",
      "Beautiful",
      "Music",
    ],
    "friendly_name": "Dining room light",
    "hs_color": [
      9,
      67.1,
    ],
    "icon": "mdi:light-recessed",
    "max_color_temp_kelvin": 6535,
    "max_mireds": 370,
    "min_color_temp_kelvin": 2702,
    "min_mireds": 153,
    "raw_state": true,
    "rgb_color": [
      255,
      109,
      83,
    ],
    "supported_color_modes": [
      "color_temp",
      "hs",
    ],
    "supported_features": 23,
    "xy_color": [
      0.591,
      0.328,
    ],
  },
  "context": {
    "id": "",
    "parent_id": null,
    "user_id": null,
  },
  "custom": {
    "active": false,
    "brightness": "brightness(100%)",
    "hexColor": "#ff6e54",
    "relativeTime": "1 year ago",
    "rgbColor": "rgba(255, 109.56074999999998, 83.89500000000001)",
    "rgbaColor": "rgba(255, 109.56074999999998, 83.89500000000001, 0.35)",
  },
  "entity_id": "light.fake_light",
  "state": "on",
}
`;
